src/extension/loadLogs.ts

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+/// <reference path="jsonSourceMap.d.ts" />
+import { execFileSync } from 'child_process';
+import * as fs from 'fs';
+import jsonMap from 'json-source-map';
+import { join } from 'path';
+import { Log } from 'sarif';
+import { eq, gt, lt } from 'semver';
+import { tmpNameSync } from 'tmp';
+import { ProgressLocation, Uri, window } from 'vscode';
+import { augmentLog, JsonMap } from '../shared';
+import { Store } from './store';
+
+export async function loadLogs(uris: Uri[], token?: { isCancellationRequested: boolean }) {
+    const logs = uris
+        .map(uri => {
+            if (token?.isCancellationRequested) return undefined;
+            try {
+                const file = fs.readFileSync(uri.fsPath, 'utf8')  // Assume scheme file.
+                    .replace(/^\uFEFF/, ''); // Trim BOM.
+                const {data: log, pointers} = jsonMap.parse(file) as { data: Log, pointers: JsonMap};
+                log._uri = uri.toString();
+                log._jsonMap = pointers;
+                return log;
+            } catch (error) {
+                window.showErrorMessage(`Failed to parse '${uri.fsPath}'`);
+                return undefined;
+            }
+        })
+        .filter(log => log) as Log[];
+    const logsNoUpgrade = [] as Log[];
+    const logsToUpgrade = [] as Log[];
+    const warnUpgradeExtension = logs.some(log => detectUpgrade(log, logsNoUpgrade, logsToUpgrade));
+    const upgrades = logsToUpgrade.length;
+    if (upgrades) {
+        await window.withProgress(
+            { location: ProgressLocation.Notification },
+            async progress => {
+                for (const [i, oldLog] of logsToUpgrade.entries()) {
+                    if (token?.isCancellationRequested) break;
+                    progress.report({
+                        message: `Upgrading ${i + 1} of ${upgrades} log${upgrades === 1 ? '' : 's'}...`,
+                        increment: 1 / upgrades * 100
+                    });
+                    await new Promise(r => setTimeout(r, 0)); // Await needed for progress to update
+                    const {fsPath} = Uri.parse(oldLog._uri);
```

Comment:
would putting the `progress.report` in a promise also solve the issue?

Reply:
trying to understand why this is needed

---
In reply to: [441212822](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r441212822) [](ancestors = 441212822)

Reply:
Assuming it did work, would you consider that a better solution? It would still require one promise, but just in a different place, and with additional nesting.

Technically progress.report() does not require an await. You could also make the case that the rest of the function body should be wrapped in a promise. I think the the result is the same.

---
In reply to: [441212822](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r441212822) [](ancestors = 441212822)

Reply:
"trying to understand why this is needed"... really a question for VS Code. I'm guessing the the `await` allows a "thread switch" which allows the renderer thread to paint.

---
In reply to: [441213031](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r441213031) [](ancestors = 441213031,441212822)

Reply:
Comments added to clarify.

---
In reply to: [441221515](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r441221515) [](ancestors = 441221515,441212822)

Reply:
IMO wrapping `progress.report` in a promise (assuming that works) seems more intentional - we are letting that action complete without being blocked by the rest of the work done in this function. Just my opinion though - either way works :)

---
In reply to: [441224736](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r441224736) [](ancestors = 441224736,441221515,441212822)